### ISSUE TITLE
Unable to download books with a library card

### ISSUE THREAD

**Issue by Qiangong2** [lorenzodifuccia/safaribooks#192]
Through my library, I get free access to Safari Books online classic. However, it makes me login with my library card, not a personal login. How can I download books this way? The normal way doesn't seem to work

**Comment by kubilayelmas** [comment]
Yes, same for me. We need an option to download without credentials or just by providing the url or something.

**Comment by ysam12345** [comment]
Can you try this:
1. create a file name "cookies.json" in root folder(same as location of safaribooks.py)
2. edit cookies.json as following format, and fill up value from your browser which already login to Safari Books online.
{
    "orm-rt": "", 
    "groot_sessionid": "", 
    "logged_in": "", 
    "orm-jwt": "", 
    "BrowserCookie": "", 
    "csrfsafari": "", 
    "salesforce_id": "", 
    "sessionid": "", 
    "csrftoken": ""
}

**Comment by ysam12345** [comment]
Can you try this:
1. Create a file name "cookies.json" in root folder(same as location of safaribooks.py).
2. Edit cookies.json as following format, and fill up values from the cookie of your browser which had already login to Safari Books online.
```
{
    "orm-rt": "", 
    "groot_sessionid": "", 
    "logged_in": "", 
    "orm-jwt": "", 
    "BrowserCookie": "", 
    "csrfsafari": "", 
    "salesforce_id": "", 
    "sessionid": "", 
    "csrftoken": ""
}
```
3. run the safaribooks.py without --cred parameter
e.g.
```
$ python3 safaribooks.py XXXXXXXXXXXXX
```
It should use your cookies.json file to authenticate.

**Comment by Qiangong2** [comment]
@ysam12345 

I just get：
```
[#] Authentication issue: unable to access profile page.                        
[!] Aborting...

**Comment by Qiangong2** [comment]
Maybe it has something to do with the fact that the url is this: https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430

(That's for the book "Absolute FreeBSD 3rd edition")

**Comment by npatel007** [comment]
i have the same issue

**Comment by npatel007** [comment]
> Maybe it has something to do with the fact that the url is this: https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430
> 
> (That's for the book "Absolute FreeBSD 3rd edition")

did you figured it out i am suspecting the same

**Comment by dan-r95** [comment]
> Can you try this:
> 
> 1. Create a file name "cookies.json" in root folder(same as location of safaribooks.py).
> 2. Edit cookies.json as following format, and fill up values from the cookie of your browser which had already login to Safari Books online.
> 
> ```
> {
>     "orm-rt": "", 
>     "groot_sessionid": "", 
>     "logged_in": "", 
>     "orm-jwt": "", 
>     "BrowserCookie": "", 
>     "csrfsafari": "", 
>     "salesforce_id": "", 
>     "sessionid": "", 
>     "csrftoken": ""
> }
> ```
> 
> 1. run the safaribooks.py without --cred parameter
>    e.g.
> 
> ```
> $ python3 safaribooks.py XXXXXXXXXXXXX
> ```
> 
> It should use your cookies.json file to authenticate.

Can confirm that works! 
On Chrome - Go to Dev Tools --> Application -> Cookies

**Comment by Qiangong2** [comment]
I still can't get it to function for some reason. What parts did you use for the cookies.json @dan-r95 ?

**Comment by dan-r95** [comment]
I pasted all values I found in the Cookie to the `cookies.json`

**Comment by Qiangong2** [comment]
The same as before
```
[#] Authentication issue: unable to access profile page.                        
[!] Aborting...  
```

**Comment by ysam12345** [comment]
> Maybe it has something to do with the fact that the url is this: https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430
> 
> (That's for the book "Absolute FreeBSD 3rd edition")

@Qiangong2 Sorry for not following this issue. I think it's possible caused by the URL is different. The URL of "Absolute FreeBSD 3rd edition" for me is https://learning.oreilly.com/library/view/absolute-freebsd-3rd/9781492069430/

Can you try to access https://learning.oreilly.com/profile/ or https://learning.oreilly.com/library/view/absolute-freebsd-3rd/9781492069430/ after login with your library card? And see if it's already login?

**Comment by Qiangong2** [comment]
Unfortunately, it doesn't let me. Maybe because it's proquest?

**Comment by ysam12345** [comment]
@Qiangong2 Yes, but I'm not sure how they connect to sfaribooksonline and handle the request. Maybe I need more information to see if I can help. 
1. Does your browser's cookies has the fields mentioned at https://github.com/lorenzodifuccia/safaribooks/issues/192#issuecomment-597660193?
2. Can you access https://learning.oreilly.com/api/v1/book/9781492069430/ after login with your library card?
3. On Chrome > Dev Tools > Network > go to https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430 > filter results by "api", do you see any url like https://learning.oreilly.com/api/v1/?
4. Can you upload a screenshot of https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430 after login with your library card?

Hope I can help!

**Comment by Qiangong2** [comment]
1. No, I do not. Mine are completely different
2. All I get is `{"detail":"Authentication credentials were not provided."}`
3. No URLs show up when searching for API
4.
![image](https://user-images.githubusercontent.com/16091790/96640727-729c1680-12d8-11eb-9ecb-d9fb7e53b519.png)

Hope this helps!

**Comment by ysam12345** [comment]
@Qiangong2 Wow, looks like this is completely another website. Can you click on the "START READING" button and see what's the URL?

**Comment by Qiangong2** [comment]
@ysam12345 Here is the URL: https://proquestpubliclibrary-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430/firstchapter

**Comment by ysam12345** [comment]
@Qiangong2 I checked the preview book html in proquest(https://proquestpubliclibrary.safaribooksonline.com/book/programming/android/9780135257555/title-page/index_html) and found it's very different from https://learning.oreilly.com/. I think if we want to download books from proquest, we might have to rewrite to code to fit the structure in proquest.  :(

**Comment by Qiangong2** [comment]
Oh no :( So it's not the same as safari books online then?

**Comment by ysam12345** [comment]
No, they looks different for me. :(

**Comment by jgarr16** [comment]
> > Maybe it has something to do with the fact that the url is this: https://proquest-safaribooksonline-com.ezproxy.spl.org/book/operating-systems-and-server-administration/freebsd/9781492069430
> > (That's for the book "Absolute FreeBSD 3rd edition")
> 
> did you figured it out i am suspecting the same

Looks like you are behind a proxy for your public library. The direct link works - https://learning.oreilly.com/library/view/absolute-freebsd-3rd/9781492069430/
